File: hosts/macos/Sources/EivizMac/TransitionCatalog.swift
import EivizMixer

enum TransitionGroup: Int, CaseIterable {
    case basic
    case wipe
    case motion
    case shader

    var title: String {
        switch self {
        case .basic: return "Basic"
        case .wipe: return "Wipe"
        case .motion: return "Motion"
        case .shader: return "Shader"
        }
    }
}

struct TransitionInfo: Identifiable {
    var id: UInt32 { kind }
    let kind: UInt32
    let label: String
    let group: TransitionGroup
    let hasDirection: Bool
    let hasDipColor: Bool
    let hasSoftness: Bool
    let hasParam: Bool
    let softnessLabel: String
    let paramLabel: String
}

enum TransitionCatalog {
    static let all: [TransitionInfo] = [
        .init(kind: EIVIZ_TRANSITION_CUT, label: "Cut", group: .basic, hasDirection: false, hasDipColor: false, hasSoftness: false, hasParam: false, softnessLabel: "", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_FADE, label: "Fade", group: .basic, hasDirection: false, hasDipColor: false, hasSoftness: false, hasParam: false, softnessLabel: "", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_DIP, label: "Dip", group: .basic, hasDirection: false, hasDipColor: true, hasSoftness: false, hasParam: false, softnessLabel: "", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_ADDITIVE, label: "Additive", group: .basic, hasDirection: false, hasDipColor: false, hasSoftness: false, hasParam: false, softnessLabel: "", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_CUSTOM, label: "Custom WGSL", group: .basic, hasDirection: false, hasDipColor: false, hasSoftness: false, hasParam: false, softnessLabel: "", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_WIPE, label: "Wipe", group: .wipe, hasDirection: true, hasDipColor: false, hasSoftness: true, hasParam: false, softnessLabel: "Edge", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_IRIS, label: "Iris", group: .wipe, hasDirection: false, hasDipColor: false, hasSoftness: true, hasParam: false, softnessLabel: "Edge", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_BLINDS, label: "Blinds", group: .wipe, hasDirection: true, hasDipColor: false, hasSoftness: true, hasParam: true, softnessLabel: "Edge", paramLabel: "Strips"),
        .init(kind: EIVIZ_TRANSITION_BARN_DOOR, label: "BarnDoor", group: .wipe, hasDirection: true, hasDipColor: false, hasSoftness: true, hasParam: false, softnessLabel: "Edge", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_CLOCK, label: "Clock", group: .wipe, hasDirection: true, hasDipColor: false, hasSoftness: true, hasParam: false, softnessLabel: "Edge", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_HEART, label: "Heart", group: .wipe, hasDirection: false, hasDipColor: false, hasSoftness: true, hasParam: false, softnessLabel: "Edge", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_DIAMOND, label: "Diamond", group: .wipe, hasDirection: false, hasDipColor: false, hasSoftness: true, hasParam: false, softnessLabel: "Edge", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_STAR, label: "Star", group: .wipe, hasDirection: false, hasDipColor: false, hasSoftness: true, hasParam: false, softnessLabel: "Edge", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_ROLLER_DOOR, label: "RollerDoor", group: .wipe, hasDirection: true, hasDipColor: false, hasSoftness: true, hasParam: false, softnessLabel: "Edge", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_SLIDE, label: "Slide", group: .motion, hasDirection: true, hasDipColor: false, hasSoftness: false, hasParam: false, softnessLabel: "", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_PUSH, label: "Push", group: .motion, hasDirection: true, hasDipColor: true, hasSoftness: false, hasParam: false, softnessLabel: "", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_ZOOM, label: "Zoom", group: .motion, hasDirection: false, hasDipColor: false, hasSoftness: false, hasParam: false, softnessLabel: "", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_CROSS_ZOOM, label: "CrossZoom", group: .motion, hasDirection: false, hasDipColor: false, hasSoftness: false, hasParam: false, softnessLabel: "", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_FLY_ROTATE, label: "FlyRotate", group: .motion, hasDirection: true, hasDipColor: false, hasSoftness: false, hasParam: true, softnessLabel: "", paramLabel: "Spin"),
        .init(kind: EIVIZ_TRANSITION_FLIP, label: "Flip", group: .motion, hasDirection: true, hasDipColor: false, hasSoftness: false, hasParam: false, softnessLabel: "", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_CUBE, label: "Cube", group: .motion, hasDirection: true, hasDipColor: false, hasSoftness: false, hasParam: false, softnessLabel: "", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_CUBE_ZOOM, label: "CubeZoom", group: .motion, hasDirection: true, hasDipColor: false, hasSoftness: false, hasParam: false, softnessLabel: "", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_MULTITASK, label: "MultiTask", group: .motion, hasDirection: true, hasDipColor: false, hasSoftness: false, hasParam: false, softnessLabel: "", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_LOREZ, label: "LoRez", group: .shader, hasDirection: false, hasDipColor: false, hasSoftness: true, hasParam: false, softnessLabel: "Pixel size", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_METAMIX, label: "MetaMix", group: .shader, hasDirection: false, hasDipColor: false, hasSoftness: false, hasParam: true, softnessLabel: "", paramLabel: "Copies"),
        .init(kind: EIVIZ_TRANSITION_TILE, label: "Tile", group: .shader, hasDirection: true, hasDipColor: false, hasSoftness: false, hasParam: true, softnessLabel: "", paramLabel: "Tiles"),
        .init(kind: EIVIZ_TRANSITION_PARTS, label: "Parts", group: .shader, hasDirection: true, hasDipColor: false, hasSoftness: false, hasParam: true, softnessLabel: "", paramLabel: "Chunks"),
        .init(kind: EIVIZ_TRANSITION_STATIC, label: "Static", group: .shader, hasDirection: false, hasDipColor: false, hasSoftness: true, hasParam: true, softnessLabel: "Edge", paramLabel: "Intensity"),
        .init(kind: EIVIZ_TRANSITION_SHIFT_RGB, label: "Shift RGB", group: .shader, hasDirection: false, hasDipColor: false, hasSoftness: false, hasParam: false, softnessLabel: "", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_DISPLACE, label: "Displace", group: .shader, hasDirection: false, hasDipColor: false, hasSoftness: false, hasParam: true, softnessLabel: "", paramLabel: "Intensity"),
        .init(kind: EIVIZ_TRANSITION_GLITCH, label: "Glitch", group: .shader, hasDirection: false, hasDipColor: true, hasSoftness: true, hasParam: true, softnessLabel: "Edge", paramLabel: "Intensity"),
        .init(kind: EIVIZ_TRANSITION_SWIRL, label: "Swirl", group: .shader, hasDirection: true, hasDipColor: false, hasSoftness: false, hasParam: true, softnessLabel: "", paramLabel: "Turns"),
        .init(kind: EIVIZ_TRANSITION_LUMA_MORPH, label: "LumaMorph", group: .shader, hasDirection: false, hasDipColor: false, hasSoftness: true, hasParam: false, softnessLabel: "Edge", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_RIPPLE, label: "Ripple", group: .shader, hasDirection: false, hasDipColor: false, hasSoftness: false, hasParam: true, softnessLabel: "", paramLabel: "Intensity"),
        .init(kind: EIVIZ_TRANSITION_GRID_DISSOLVE, label: "GridDissolve", group: .shader, hasDirection: false, hasDipColor: false, hasSoftness: true, hasParam: true, softnessLabel: "Edge", paramLabel: "Cells"),
        .init(kind: EIVIZ_TRANSITION_POLAR, label: "Polar", group: .shader, hasDirection: true, hasDipColor: false, hasSoftness: false, hasParam: false, softnessLabel: "", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_KALEIDOSCOPE, label: "Kaleidoscope", group: .shader, hasDirection: false, hasDipColor: false, hasSoftness: false, hasParam: true, softnessLabel: "", paramLabel: "Segments"),
        .init(kind: EIVIZ_TRANSITION_PAGE_CURL, label: "PageCurl", group: .shader, hasDirection: true, hasDipColor: false, hasSoftness: false, hasParam: false, softnessLabel: "", paramLabel: ""),
        .init(kind: EIVIZ_TRANSITION_FILM_BURN, label: "FilmBurn", group: .shader, hasDirection: false, hasDipColor: true, hasSoftness: true, hasParam: true, softnessLabel: "Edge", paramLabel: "Intensity"),
        .init(kind: EIVIZ_TRANSITION_ZOOM_BLUR, label: "ZoomBlur", group: .shader, hasDirection: false, hasDipColor: false, hasSoftness: false, hasParam: true, softnessLabel: "", paramLabel: "Intensity"),
        .init(kind: EIVIZ_TRANSITION_PIXEL_SORT, label: "PixelSort", group: .shader, hasDirection: true, hasDipColor: false, hasSoftness: true, hasParam: true, softnessLabel: "Threshold", paramLabel: "Span"),
        .init(kind: EIVIZ_TRANSITION_DATAMOSH, label: "Datamosh", group: .shader, hasDirection: true, hasDipColor: false, hasSoftness: false, hasParam: true, softnessLabel: "", paramLabel: "Intensity"),
        .init(kind: EIVIZ_TRANSITION_VISUAL_DISSOLVE, label: "VisualDissolve", group: .shader, hasDirection: false, hasDipColor: false, hasSoftness: true, hasParam: true, softnessLabel: "Edge", paramLabel: "Flow"),
        .init(kind: EIVIZ_TRANSITION_OPTICAL_FLOW, label: "OpticalFlow", group: .shader, hasDirection: false, hasDipColor: false, hasSoftness: false, hasParam: true, softnessLabel: "", paramLabel: "Amount"),
        .init(kind: EIVIZ_TRANSITION_BLOOM, label: "Bloom", group: .shader, hasDirection: false, hasDipColor: false, hasSoftness: true, hasParam: true, softnessLabel: "Threshold", paramLabel: "Intensity"),
    ]

    static func info(_ kind: UInt32) -> TransitionInfo {
        all.first { $0.kind == kind } ?? all[1]
    }

    static func label(_ kind: UInt32) -> String {
        kind == EIVIZ_TRANSITION_STINGER ? "Stinger" : info(kind).label
    }

    static func items(in group: TransitionGroup) -> [TransitionInfo] {
        all.filter { $0.group == group }
    }

    static func showsSoftness(_ kind: UInt32) -> Bool {
        let label = info(kind).softnessLabel
        return !label.isEmpty && label != "Edge"
    }

    static func defaultSoftness(_ kind: UInt32) -> Float {
        switch kind {
        case EIVIZ_TRANSITION_PIXEL_SORT: return 0.4
        case EIVIZ_TRANSITION_BLOOM: return 0.45
        default: return 0.02
        }
    }

    static func defaultParam(_ kind: UInt32) -> Float {
        switch kind {
        case EIVIZ_TRANSITION_PIXEL_SORT: return 0.25
        case EIVIZ_TRANSITION_DATAMOSH: return 1
        case EIVIZ_TRANSITION_METAMIX: return 8
        case EIVIZ_TRANSITION_TILE: return 8
        case EIVIZ_TRANSITION_PARTS: return 6
        case EIVIZ_TRANSITION_VISUAL_DISSOLVE: return 0.42
        case EIVIZ_TRANSITION_OPTICAL_FLOW: return 1
        case EIVIZ_TRANSITION_BLOOM: return 0.85
        default: return 0
        }
    }

    static func defaultDurationValue(_ kind: UInt32) -> UInt32 {
        kind == EIVIZ_TRANSITION_PIXEL_SORT ? 45 : 0
    }

    static func defaultDirection(_ kind: UInt32) -> UInt32? {
        kind == EIVIZ_TRANSITION_PIXEL_SORT ? 3 : nil
    }

    static func applyKindDefaults(_ preset: inout TransitionPreset) {
        if preset.kind == EIVIZ_TRANSITION_PIXEL_SORT {
            let oldPair = abs(preset.softness - 0.3) < 0.0005 && abs(preset.param - 0.4) < 0.0005
            if oldPair {
                preset.softness = defaultSoftness(preset.kind)
                preset.param = defaultParam(preset.kind)
                if preset.durationUnit == 0 && preset.durationValue == 30 {
                    preset.durationValue = defaultDurationValue(preset.kind)
                }
                if preset.direction == 0 {
                    preset.direction = defaultDirection(preset.kind) ?? 3
                }
            }
            if preset.softness <= 0.021 { preset.softness = defaultSoftness(preset.kind) }
            if preset.param <= 0 { preset.param = defaultParam(preset.kind) }
        }
    }
}

File: hosts/win32/TransitionCatalog.cs
namespace Eiviz.Host.Interop;

internal enum TransitionGroup
{
    Basic,
    Wipe,
    Motion,
    Shader
}

internal readonly record struct TransitionInfo(
    uint Kind,
    string Label,
    TransitionGroup Group,
    bool HasDirection,
    bool HasDipColor,
    bool HasSoftness,
    bool HasParam,
    string SoftnessLabel,
    string ParamLabel);

internal static class TransitionCatalog
{
    internal static readonly TransitionInfo[] All =
    [
        new(MixerNative.TransitionCut, "Cut", TransitionGroup.Basic, false, false, false, false, "", ""),
        new(MixerNative.TransitionFade, "Fade", TransitionGroup.Basic, false, false, false, false, "", ""),
        new(MixerNative.TransitionDip, "Dip", TransitionGroup.Basic, false, true, false, false, "", ""),
        new(MixerNative.TransitionAdditive, "Additive", TransitionGroup.Basic, false, false, false, false, "", ""),
        new(MixerNative.TransitionCustom, "Custom WGSL", TransitionGroup.Basic, false, false, false, false, "", ""),
        new(MixerNative.TransitionWipe, "Wipe", TransitionGroup.Wipe, true, false, true, false, "Edge", ""),
        new(MixerNative.TransitionIris, "Iris", TransitionGroup.Wipe, false, false, true, false, "Edge", ""),
        new(MixerNative.TransitionBlinds, "Blinds", TransitionGroup.Wipe, true, false, true, true, "Edge", "Strips"),
        new(MixerNative.TransitionBarnDoor, "BarnDoor", TransitionGroup.Wipe, true, false, true, false, "Edge", ""),
        new(MixerNative.TransitionClock, "Clock", TransitionGroup.Wipe, true, false, true, false, "Edge", ""),
        new(MixerNative.TransitionHeart, "Heart", TransitionGroup.Wipe, false, false, true, false, "Edge", ""),
        new(MixerNative.TransitionDiamond, "Diamond", TransitionGroup.Wipe, false, false, true, false, "Edge", ""),
        new(MixerNative.TransitionStar, "Star", TransitionGroup.Wipe, false, false, true, false, "Edge", ""),
        new(MixerNative.TransitionRollerDoor, "RollerDoor", TransitionGroup.Wipe, true, false, true, false, "Edge", ""),
        new(MixerNative.TransitionSlide, "Slide", TransitionGroup.Motion, true, false, false, false, "", ""),
        new(MixerNative.TransitionPush, "Push", TransitionGroup.Motion, true, true, false, false, "", ""),
        new(MixerNative.TransitionZoom, "Zoom", TransitionGroup.Motion, false, false, false, false, "", ""),
        new(MixerNative.TransitionCrossZoom, "CrossZoom", TransitionGroup.Motion, false, false, false, false, "", ""),
        new(MixerNative.TransitionFlyRotate, "FlyRotate", TransitionGroup.Motion, true, false, false, true, "", "Spin"),
        new(MixerNative.TransitionFlip, "Flip", TransitionGroup.Motion, true, false, false, false, "", ""),
        new(MixerNative.TransitionCube, "Cube", TransitionGroup.Motion, true, false, false, false, "", ""),
        new(MixerNative.TransitionCubeZoom, "CubeZoom", TransitionGroup.Motion, true, false, false, false, "", ""),
        new(MixerNative.TransitionMultitask, "MultiTask", TransitionGroup.Motion, true, false, false, false, "", ""),
        new(MixerNative.TransitionLorez, "LoRez", TransitionGroup.Shader, false, false, true, false, "Pixel size", ""),
        new(MixerNative.TransitionMetamix, "MetaMix", TransitionGroup.Shader, false, false, false, true, "", "Copies"),
        new(MixerNative.TransitionTile, "Tile", TransitionGroup.Shader, true, false, false, true, "", "Tiles"),
        new(MixerNative.TransitionParts, "Parts", TransitionGroup.Shader, true, false, false, true, "", "Chunks"),
        new(MixerNative.TransitionStatic, "Static", TransitionGroup.Shader, false, false, true, true, "Edge", "Intensity"),
        new(MixerNative.TransitionShiftRgb, "Shift RGB", TransitionGroup.Shader, false, false, false, false, "", ""),
        new(MixerNative.TransitionDisplace, "Displace", TransitionGroup.Shader, false, false, false, true, "", "Intensity"),
        new(MixerNative.TransitionGlitch, "Glitch", TransitionGroup.Shader, false, true, true, true, "Edge", "Intensity"),
        new(MixerNative.TransitionSwirl, "Swirl", TransitionGroup.Shader, true, false, false, true, "", "Turns"),
        new(MixerNative.TransitionLumaMorph, "LumaMorph", TransitionGroup.Shader, false, false, true, false, "Edge", ""),
        new(MixerNative.TransitionRipple, "Ripple", TransitionGroup.Shader, false, false, false, true, "", "Intensity"),
        new(MixerNative.TransitionGridDissolve, "GridDissolve", TransitionGroup.Shader, false, false, true, true, "Edge", "Cells"),
        new(MixerNative.TransitionPolar, "Polar", TransitionGroup.Shader, true, false, false, false, "", ""),
        new(MixerNative.TransitionKaleidoscope, "Kaleidoscope", TransitionGroup.Shader, false, false, false, true, "", "Segments"),
        new(MixerNative.TransitionPageCurl, "PageCurl", TransitionGroup.Shader, true, false, false, false, "", ""),
        new(MixerNative.TransitionFilmBurn, "FilmBurn", TransitionGroup.Shader, false, true, true, true, "Edge", "Intensity"),
        new(MixerNative.TransitionZoomBlur, "ZoomBlur", TransitionGroup.Shader, false, false, false, true, "", "Intensity"),
        new(MixerNative.TransitionPixelSort, "PixelSort", TransitionGroup.Shader, true, false, true, true, "Threshold", "Span"),
        new(MixerNative.TransitionDatamosh, "Datamosh", TransitionGroup.Shader, true, false, false, true, "", "Intensity"),
        new(MixerNative.TransitionVisualDissolve, "VisualDissolve", TransitionGroup.Shader, false, false, true, true, "Edge", "Flow"),
        new(MixerNative.TransitionOpticalFlow, "OpticalFlow", TransitionGroup.Shader, false, false, false, true, "", "Amount"),
        new(MixerNative.TransitionBloom, "Bloom", TransitionGroup.Shader, false, false, true, true, "Threshold", "Intensity"),
    ];

    internal static TransitionInfo Info(uint kind) =>
        All.FirstOrDefault(item => item.Kind == kind, All[1]);

    internal static string Label(uint kind) => kind switch
    {
        MixerNative.TransitionStinger => "Stinger",
        _ => Info(kind).Label
    };

    internal static string GroupName(TransitionGroup group) => group switch
    {
        TransitionGroup.Wipe => "Wipe",
        TransitionGroup.Motion => "Motion",
        TransitionGroup.Shader => "Shader",
        _ => "Basic"
    };

    internal static float DefaultSoftness(uint kind) => kind switch
    {
        MixerNative.TransitionPixelSort => 0.4f,
        MixerNative.TransitionBloom => 0.45f,
        _ => 0.02f
    };

    internal static float DefaultParam(uint kind) => kind switch
    {
        MixerNative.TransitionPixelSort => 0.25f,
        MixerNative.TransitionDatamosh => 1f,
        MixerNative.TransitionMetamix => 8f,
        MixerNative.TransitionTile => 8f,
        MixerNative.TransitionParts => 6f,
        MixerNative.TransitionVisualDissolve => 0.42f,
        MixerNative.TransitionOpticalFlow => 1f,
        MixerNative.TransitionBloom => 0.85f,
        _ => 0f
    };

    internal static uint DefaultDurationValue(uint kind) => kind switch
    {
        MixerNative.TransitionPixelSort => 45,
        _ => 0
    };

    internal static uint? DefaultDirection(uint kind) => kind switch
    {
        MixerNative.TransitionPixelSort => 3u,
        _ => null
    };

    internal static bool ShowsSoftness(uint kind)
    {
        var label = Info(kind).SoftnessLabel;
        return !string.IsNullOrEmpty(label) && label != "Edge";
    }

    internal static void ApplyKindDefaults(TransitionPreset preset)
    {
        if (preset.Kind == MixerNative.TransitionPixelSort)
        {
            var oldPair = Math.Abs(preset.Softness - 0.3f) < 0.0005f && Math.Abs(preset.Param - 0.4f) < 0.0005f;
            if (oldPair)
            {
                preset.Softness = DefaultSoftness(preset.Kind);
                preset.Param = DefaultParam(preset.Kind);
                if (preset.DurationUnit == 0 && preset.DurationValue == 30)
                    preset.DurationValue = DefaultDurationValue(preset.Kind);
                if (preset.Direction == 0)
                    preset.Direction = DefaultDirection(preset.Kind) ?? 3u;
            }
            if (preset.Softness <= 0.021f)
                preset.Softness = DefaultSoftness(preset.Kind);
            if (preset.Param <= 0f)
                preset.Param = DefaultParam(preset.Kind);
        }
    }
}
